app/services/samples/metadata/update_service.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module Samples
+  module Metadata
+    # Service used to Update Samples::Metadata
+    class UpdateService < BaseService
+      SampleMetadataUpdateError = Class.new(StandardError)
+      attr_accessor :sample, :metadata, :analysis_id
+
+      def initialize(project, sample, user = nil, params = {})
+        super(user, params)
+        @project = project
+        @sample = sample
+        @metadata = params['metadata']
+        @analysis_id = params['analysis_id']
+      end
+
+      def execute # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/MethodLength
+        authorize! sample.project, to: :update_sample?
+        metadata_fields_update_status = { updated: [], not_updated: [] }
+
+        if @project.id != @sample.project.id
+          raise SampleMetadataUpdateError,
+                I18n.t('services.samples.metadata.sample_does_not_belong_to_project', sample_name: @sample.name,
+                                                                                      project_name: @project.name)
+        end
+
+        if @metadata.nil? || @metadata == {}
+          raise SampleMetadataUpdateError,
+                I18n.t('services.samples.metadata.empty_metadata', sample_name: @sample.name)
+        else
+          # Without transforming keys, issues with overwritting can occur and multiples of the same key can appear
+          @metadata = @metadata.transform_keys(&:to_s)
+          @metadata.each do |key, value|
+            if value.blank?
+              if @sample['metadata'].key?(key)
+                @sample['metadata'].delete(key)
+                @sample['metadata_provenance'].delete(key)
+                metadata_fields_update_status[:updated].append(key)
+              end
+            else
+              metadata_updated = update_metadata(key, value)
+              if metadata_updated
+                metadata_fields_update_status[:updated].append(key)
+              else
+                metadata_fields_update_status[:not_updated].append(key)
+              end
+            end
+          end
+          @sample.update(id: @sample.id)
```

Comment:
**nitpick (blocking):** What is this link supposed to do? if you just want to persist changes to the metadata after the if then just call `@sample.save`.

Reply:
Fixed in b78174939f9a80456d3eb8fbc8342885608f718e

Reply:
Fixed in b78174939f9a80456d3eb8fbc8342885608f718e
Moved `key_transformation` and checking `not_updated_metadata_fields` to their own private methods too, was able to get rid of `cyclomatic` and `perceived complexity` disables.